source/util/parse_number.h

```diff
@@ -0,0 +1,122 @@
+  bool isSigned;      // This is only significant if type_class is integral.
+  IdTypeClass type_class;
+};
+
+// Default equality operator for IdType. Tests if all members are the same.
+inline bool operator==(const IdType& first, const IdType& second) {
+  return (first.bitwidth == second.bitwidth) &&
+         (first.isSigned == second.isSigned) &&
+         (first.type_class == second.type_class);
+}
+
+// Tests whether any member of the IdTypes do not match.
+inline bool operator!=(const IdType& first, const IdType& second) {
+  return !(first == second);
+}
+
+// Returns true if the type is a scalar integer type.
+inline bool isScalarIntegral(const IdType& type) {
+  return type.type_class == IdTypeClass::kScalarIntegerType;
+}
+
+// Returns true if the type is a scalar floating point type.
+inline bool isScalarFloating(const IdType& type) {
+  return type.type_class == IdTypeClass::kScalarFloatType;
+}
+
+// Returns the number of bits in the type.
+// This is only valid for bottom, scalar integer, and scalar floating
+// classes.  For bottom, assume 32 bits.
+inline int assumedBitWidth(const IdType& type) {
+  switch (type.type_class) {
+    case IdTypeClass::kBottom:
+      return 32;
+    case IdTypeClass::kScalarIntegerType:
+    case IdTypeClass::kScalarFloatType:
+      return type.bitwidth;
+    default:
+      break;
+  }
+  // We don't care about this case.
+  return 0;
+}
+
+enum class EncodeNumberResult {
+  kSuccess = 0,
+  kUnsupported,
+  kInvalidUsage,
+  kInvalidText,
+};
+
+// Parses a numeric value of a given type from the given text.  The number
+// should take up the entire string, and should be within bounds for the
+// target type.  On success, returns SPV_SUCCESS and populates the object
+// referenced by value_pointer. On failure, returns the given error code,
+// and emits a diagnostic if that error code is not SPV_FAILED_MATCH.
+template <typename T>
+bool ParseNumber(const char* text, T* value_pointer);
+
+// Appends the given integer literal to the given instruction.
+// Returns SPV_SUCCESS if the value was correctly parsed.  Otherwise
+// returns the given error code, and emits a diagnostic if that error
+// code is not SPV_FAILED_MATCH.
+// Integers up to 64 bits are supported.
+std::tuple<EncodeNumberResult, std::string> EncodeIntegerNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
+
+// Appends the given floating point literal to the given instruction.
+// Returns SPV_SUCCESS if the value was correctly parsed.  Otherwise
+// returns the given error code, and emits a diagnostic if that error
+// code is not SPV_FAILED_MATCH.
+// Only 32 and 64 bit floating point numbers are supported.
+std::tuple<EncodeNumberResult, std::string> EncodeFloatNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
+
+std::tuple<EncodeNumberResult, std::string> EncodeNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
```

Comment:
Hmm, I think `std::function<void(uint32_t)>` is not that long. Actually I feel not necessary to introduce a new name here.